matching-core/src/main/java/com/graphhopper/matching/GPXExtension.java

```diff
@@ -17,36 +17,50 @@
  */
 package com.graphhopper.matching;
 
+import com.graphhopper.routing.VirtualEdgeIteratorState;
 import com.graphhopper.storage.index.QueryResult;
 import com.graphhopper.util.GPXEntry;
 
 /**
+ * Represents a map matching candidate.
  *
  * @author Peter Karich
+ * @author kodonnell
+ * @author Stefan Holder
  */
 public class GPXExtension {
-    final GPXEntry entry;
-    final QueryResult queryResult;
-    final int gpxListIndex;
+    public final GPXEntry entry;
+    public final QueryResult queryResult;
+    public final boolean isDirected;
+    public final VirtualEdgeIteratorState incomingVirtualEdge;
+    public final VirtualEdgeIteratorState outgoingVirtualEdge;
 
-    public GPXExtension(GPXEntry entry, QueryResult queryResult, int gpxListIndex) {
+    public GPXExtension(GPXEntry entry, QueryResult queryResult) {
         this.entry = entry;
         this.queryResult = queryResult;
-        this.gpxListIndex = gpxListIndex;
+        this.isDirected = false;
+        this.incomingVirtualEdge = null;
+        this.outgoingVirtualEdge = null;
+    }
+    
+    public GPXExtension(GPXEntry entry, QueryResult queryResult,
+                        VirtualEdgeIteratorState incomingVirtualEdge,
+                        VirtualEdgeIteratorState outgoingVirtualEdge) {
+        this.entry = entry;
+        this.queryResult = queryResult;
+        this.isDirected = true;
+        this.incomingVirtualEdge = incomingVirtualEdge;
+        this.outgoingVirtualEdge = outgoingVirtualEdge;
     }
 
     @Override
     public String toString() {
-        return "entry:" + entry
-                + ", query distance:" + queryResult.getQueryDistance()
-                + ", gpxListIndex:" + gpxListIndex;
-    }
-
-    public QueryResult getQueryResult() {
```

Comment:
Good point - I prefer this in terms of code cleanliness/development (though I'm not sure if it's good/bad practice etc.).